instances.ml

```diff
@@ -0,0 +1,51 @@
+open Imp.Any
+open Classes
+
+implicit module Pair {A: Any} {B: Any}
+:  Product
+  with type t = A.t * B.t
+  and type a = A.t
+  and type b = B.t
+= struct
+  type t = A.t * B.t
+  type a = A.t
+  type b = B.t
+  let construct a b = .< .~a, .~b >.
+  let deconstruct p = .< fst .~p >., .< snd .~p >.
+end
+
+implicit module Option {A: Any}
+: Sum
+  with type t = A.t option
+  and type a = A.t
+  and type b = unit
+= struct
+  type t = A.t option
+  type a = A.t
+  type b = unit
+  let construct_a a = .< Some .~a >.
+  let construct_b _ = .< None >.
+  let match_ fa fb x = .<
+    match .~x with
+    | Some a -> .~(fa .< a >.)
+    | None -> .~(fb .< () >.)
+  >.
+end
+
+implicit module List {A: Any}
+: Sum
+  with type t = A.t list
+  and type a = unit
+  and type b = A.t * A.t list
+= struct
+  type t = A.t list
+  type a = unit
+  type b = A.t * A.t list
+  let construct_a _ = .< [] >.
+  let construct_b p = .< let (x, xs) = .~p in x :: xs >.
+  let match_ fa fb (x: A.t list code) = .<
+    match .~x with
+    | [] -> .~(fa .< () >.)
+    | x :: xs -> .~(fb .< (x, xs) >.)
+  >.
+end
```

Comment:
This "just works", even though `list` is recursive. But does it make sense? Encoding a list as a sum type seems like it could lead to unnecessarily stupid / slow behaviour sometimes.

Reply:
I think it makes sense, even if it's not the fastest approach.  Eventually it might be worth using a better representation than plain `code`, e.g. a data type that remembers more of the structure of the data it represents:

```ocaml
type 'a cody =
    Code of 'a code
 | Pair : 'a cody * 'b cody -> ('a * 'b) cody
 | SumL : 'a cody -> ('a, 'b) either cody
 | SumR : 'b cody -> ('a, 'b) either cody
```

Then you can write

```ocaml
let construct : type a b. a cody * b cody -> (a * b) cody =
  fun a b -> Pair (a, b)

let deconstruct : type a b. (a * b) cody -> a cody * b cody = function
  | Code c -> Code .< fst .~c>., Code .< snd .~c >.
  | Pair (a, b) -> a, b
```

etc., and it should be possible to avoid a lot of the cost of the encoding.